File: ProgrammersAlgorithm/JS/Level_2/GroundPicking.js
function solution(land) {
  var answer = 0;
  let dp = Array.from(Array(land.length), () => Array(4).fill(0));

  for(let i=0;i<4;i++) {
    dp[0][i] = land[0][i];
  }

  for(let i=1;i<land.length;i++) {
    dp[i][0] = Math.max(dp[i-1][1], dp[i-1][2], dp[i-1][3]) + land[i][0];
    dp[i][1] = Math.max(dp[i-1][0], dp[i-1][2], dp[i-1][3]) + land[i][1];
    dp[i][2] = Math.max(dp[i-1][0], dp[i-1][1], dp[i-1][3]) + land[i][2];
    dp[i][3] = Math.max(dp[i-1][0], dp[i-1][1], dp[i-1][2]) + land[i][3];
  }

  return Math.max(...dp[land.length-1]);
}

File: ProgrammersAlgorithm/Cpp/Level_2/GroundPicking.cpp
#include <iostream>
#include <vector>
using namespace std;

int getMax(vector<int> land, int index)
{

  int max = -999;

  for (int i = 0; i < land.size(); i++)
  {
    if (i != index)
    {
      if (land[i] > max)
        max = land[i];
    }
  }

  return max;
}

int solution(vector<vector<int>> land)
{
  int answer = 0;

  vector<vector<int>> dp(land.size(), vector<int>(4, 0));

  dp[0][0] = land[0][0];
  dp[0][1] = land[0][1];
  dp[0][2] = land[0][2];
  dp[0][3] = land[0][3];

  for (int i = 1; i < land.size(); i++)
  {
    for (int j = 0; j < 4; j++)
    {
      int maxValue = getMax(dp[i - 1], j);
      dp[i][j] = land[i][j] + maxValue;
    }
  }

  for (int i = 0; i < 4; i++)
  {
    int temp = dp[land.size() - 1][i];
    if (temp > answer)
      answer = temp;
  }

  return answer;
}